packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,356 @@
+        });
+    }
+
+}
+
+/**
+ * Just in case we need to support a bit more sophisticated state transitions than having `online` and `offline`.
+ * For instance, `pending`, `reconnecting`, etc...
+ */
+export class ConnectionStateMachine implements ConnectionStatus {
+
+    private static readonly MAX_HISTORY = 100;
+
+    public readonly health: number;
+
+    constructor(
+        private readonly props: { readonly threshold: number },
+        public readonly state: ConnectionState = ConnectionState.ONLINE,
+        private readonly history: boolean[] = []) {
+
+        if (this.state === ConnectionState.OFFLINE) {
+            this.health = 0;
+        } else {
+            this.health = this.history.length === 0 ? 100 : Math.round((this.history.filter(success => success).length / this.history.length) * 100);
+        }
+    }
+
+    next(success: boolean): ConnectionStateMachine {
+        const newHistory = this.updateHistory(success);
+        // Initial optimism.
+        let online = true;
+        if (newHistory.length > this.props.threshold) {
+            online = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, online ? ConnectionState.ONLINE : ConnectionState.OFFLINE, newHistory);
+    }
+
+    protected updateHistory(success: boolean) {
+        const updated = [...this.history, success];
+        if (updated.length > ConnectionStateMachine.MAX_HISTORY) {
+            updated.shift();
+        }
+        return updated;
+    }
+
+}
+
+@injectable()
+export class ConnectionStatusStatusBarContribution extends DefaultFrontendApplicationContribution {
+
+    private idle: boolean;
+
+    constructor(
+        @inject(ConnectionStatusService) protected readonly connectionStatusService: ConnectionStatusService,
+        @inject(FrontendIdleService) protected idleService: FrontendIdleService,
+        @inject(StatusBar) protected statusBar: StatusBar
+    ) {
+        super();
+        this.connectionStatusService.onStatusChange(event => this.onStatusChange(event));
+        this.idle = this.idleService.idle;
+        this.idleService.onStatusChange(idle => {
+            if (this.idle !== idle) {
+                this.idle = idle;
+                if (this.idle) {
+                    this.statusBar.removeElement('connection-status');
+                    this.statusBar.setElement('connection-status', {
+                        alignment: StatusBarAlignment.RIGHT,
+                        text: '$(bed)',
+                        priority: 0,
+                        tooltip: 'Idle'
+                    });
+                } else {
+                    this.onStatusChange(this.connectionStatusService.currentState);
+                }
+            }
+        });
+    }
+
+    protected onStatusChange(event: ConnectionStatus) {
```

Comment:
I think this method should be called both on `connectionStatusService.onStatusChange` and `idleService.onStatusChange`, remove `connection-status`, retrieve current values using getters and then render based on them

Reply:
I think the idle status should be merged into the connection status, new enumeration type (IDLE) has to be introduced.

Reply:
fine with me, I was mostly concerned about keeping the logic of rendering `connection-status` in one place